saleor/graphql/meta/mutations.py

```diff
@@ -399,3 +402,102 @@ def perform_mutation(cls, _root, info, **data):
                 instance.delete_value_from_private_metadata(key)
             _save_instance(instance, "private_metadata")
         return cls.success_response(instance)
+
+
+class BaseMutationWithMetadata(BaseMutation):
+    class Arguments:
+        private_metadata = NonNullList(
+            MetadataInput,
+            description=(
+                "Fields required to update the object's private metadata." + ADDED_IN_38
+            ),
+            required=False,
+        )
+        metadata = NonNullList(
+            MetadataInput,
+            description=(
+                "Fields required to update the object's metadata." + ADDED_IN_38
+            ),
+            required=False,
+        )
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def __init_subclass_with_meta__(
+        cls,
+        metadata_permissions_map_key=None,
+        _meta=None,
+        **kwargs,
+    ):
+        if not _meta:
+            _meta = MutationOptions(cls)
+
+        _meta.metadata_permissions_map_key = metadata_permissions_map_key
+        super().__init_subclass_with_meta__(_meta=_meta, **kwargs)
+
+    @classmethod
+    def validate_metadata_keys(cls, metadata_list: List[dict]):
+        if metadata_contains_empty_key(metadata_list):
+            raise ValidationError(
+                {
+                    "input": ValidationError(
+                        "Metadata key cannot be empty.",
+                        code=MetadataErrorCode.REQUIRED.value,
+                    )
+                }
+            )
+
+    @classmethod
+    def check_metadata_permissions(cls, type_name, info, object_id, private=False):
+        if private:
+            meta_permission = PRIVATE_META_PERMISSION_MAP.get(type_name)
+            map_name = "PRIVATE_META_PERMISSION_MAP"
+        else:
+            meta_permission = PUBLIC_META_PERMISSION_MAP.get(type_name)
+            map_name = "PUBLIC_META_PERMISSION_MAP"
+
+        if not meta_permission:
+            raise NotImplementedError(
+                f"Couldn't resolve permission to item type: {type_name}. "
+                f"Make sure that type exists inside {map_name}."
+            )
+
+        if not cls.check_permissions(info.context, meta_permission(info, object_id)):
+            metadata_type = "private metadata" if private else "metadata"
+            message = f"You do not have permission to update {metadata_type}."
+            raise PermissionDenied(message)
+
+    @classmethod
+    def mutate(cls, root, info, **data):
```

Comment:
I would recommend keeping this mutation as tight as possible. We should place here only a custom logic related to metadata permission, but we also should run `result = super().mutate(root, info, **data)`. Thanks to this, any updates that we will add to baseMutation, will apply to this mutation also. Without this, we need to put additional attention to maintaining this part of the code and keeping it up to date.